MSVC warnings when using std::fill in generated messages with types smaller than int

## Bug report

Follow up from ros2/rcl_interfaces#42.

**Required Info:**

- Operating System:
  - Windows 10
- Installation type:
  - from source
- Version or commit hash:
  - 1653fa254400ef1b364842b7d24c9db4032bd57a (master)
- DDS implementation:
  - Fast-RTPS and RTI Connext

#### Steps to reproduce issue

Either revert the patch introduced by #307 and build [test_msgs](https://github.com/ros2/rcl_interfaces), or for a [SSCCE](http://sscce.org/):

1. Copy the following program into Microsoft Visual Studio:

```c++
#include <vector>
int main(void)
{
	std::vector<char> vec(10);
	std::fill(vec.begin(), vec.end(), 0);
	return 0;
}
```

2. Turn on warning level W4 or higher:

Project -> Properties -> C/C++ -> Warning Level

3. Build

#### Expected behavior

Similar to `g++`, we would expect the program to compile fine without warnings.

#### Actual behavior

Compiler warnings:

```
>c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.12.25827\include\xutility(2888): warning C4244: '=': conversion from 'const int' to 'char', possible loss of data
1>c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.12.25827\include\xutility(2902): note: see reference to function template instantiation 'void std::_Fill_unchecked1<_FwdIt,_Ty>(_FwdIt,_FwdIt,const _Ty &,std::false_type)' being compiled
1>        with
1>        [
1>            _FwdIt=char *,
1>            _Ty=int
1>        ]
1>c:\program files (x86)\microsoft visual studio\2017\community\vc\tools\msvc\14.12.25827\include\xutility(2910): note: see reference to function template instantiation 'void std::_Fill_unchecked<char*,_Ty>(_FwdIt,_FwdIt,const _Ty &)' being compiled
1>        with
1>        [
1>            _Ty=int,
1>            _FwdIt=char *
1>        ]
1>c:\users\osrf\source\repos\project1\project1\main.cpp(5): note: see reference to function template instantiation 'void std::fill<std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>,int>(_FwdIt,_FwdIt,const int &)' being compiled
1>        with
1>        [
1>            _Ty=char,
1>            _FwdIt=std::_Vector_iterator<std::_Vector_val<std::_Simple_types<char>>>
1>        ]
```

#### Additional information

See TODOs in the rosidl_generator_cpp with a workaround:

https://github.com/ros2/rosidl/blob/6c8e2deaa22e7be78b827820d3471be046a536a4/rosidl_generator_cpp/resource/msg__struct.hpp.em#L103-L104

https://github.com/ros2/rosidl/blob/6c8e2deaa22e7be78b827820d3471be046a536a4/rosidl_generator_cpp/resource/msg__struct.hpp.em#L124-L125

See ros2/rcl_interfaces#42 for more information.

--- a/rosidl_generator_cpp/resource/msg__struct.hpp.em
+++ b/rosidl_generator_cpp/resource/msg__struct.hpp.em
@@ -5,6 +5,8 @@ from rosidl_generator_cpp import escape_string
 from rosidl_generator_cpp import escape_wstring
 from rosidl_generator_cpp import msg_type_to_cpp
 from rosidl_generator_cpp import MSG_TYPE_TO_CPP
+from rosidl_generator_cpp import generate_zero_string
+from rosidl_generator_cpp import generate_default_string
 from rosidl_parser.definition import AbstractNestedType
 from rosidl_parser.definition import AbstractString
 from rosidl_parser.definition import AbstractWString
@@ -108,48 +110,6 @@ struct @(message.structure.namespaced_type.name)_
 # http://design.ros2.org/articles/generated_interfaces_cpp.html#constructors
 # for a detailed explanation of the different _init parameters.
 init_list, alloc_list, member_list = create_init_alloc_and_member_lists(message)
-
-def generate_default_string(membset):
-    from rosidl_generator_cpp import msg_type_only_to_cpp
-    from rosidl_generator_cpp import msg_type_to_cpp
-    strlist = []
-    for member in membset.members:
-        if member.default_value is not None:
-            if member.num_prealloc > 0:
-                strlist.append('this->%s.resize(%d);' % (member.name, member.num_prealloc))
-            if isinstance(member.default_value, list):
-                if all(v == member.default_value[0] for v in member.default_value):
-                    # Specifying type for std::fill because of MSVC 14.12 warning about casting 'const int' to smaller types (C4244)
-                    # For more info, see https://github.com/ros2/rosidl/issues/309
-                    # TODO(jacobperron): Investigate reason for build warnings on Windows
-                    # TODO(jacobperron): Write test case for this path of execution
-                    strlist.append('std::fill<typename %s::iterator, %s>(this->%s.begin(), this->%s.end(), %s);' % (msg_type_to_cpp(member.type), msg_type_only_to_cpp(member.type), member.name, member.name, member.default_value[0]))
-                else:
-                    for index, val in enumerate(member.default_value):
-                        strlist.append('this->%s[%d] = %s;' % (member.name, index, val))
-            else:
-                strlist.append('this->%s = %s;' % (member.name, member.default_value))
-
-    return strlist
-
-def generate_zero_string(membset, fill_args):
-    from rosidl_generator_cpp import msg_type_only_to_cpp
-    from rosidl_generator_cpp import msg_type_to_cpp
-    strlist = []
-    for member in membset.members:
-        if isinstance(member.zero_value, list):
-            if member.num_prealloc > 0:
-                strlist.append('this->%s.resize(%d);' % (member.name, member.num_prealloc))
-            if member.zero_need_array_override:
-                strlist.append('this->%s.fill(%s{%s});' % (member.name, msg_type_only_to_cpp(member.type), fill_args))
-            else:
-                # Specifying type for std::fill because of MSVC 14.12 warning about casting 'const int' to smaller types (C4244)
-                # For more info, see https://github.com/ros2/rosidl/issues/309
-                # TODO(jacobperron): Investigate reason for build warnings on Windows
-                strlist.append('std::fill<typename %s::iterator, %s>(this->%s.begin(), this->%s.end(), %s);' % (msg_type_to_cpp(member.type), msg_type_only_to_cpp(member.type), member.name, member.name, member.zero_value[0]))
-        else:
-            strlist.append('this->%s = %s;' % (member.name, member.zero_value))
-    return strlist
 }@
   explicit @(message.structure.namespaced_type.name)_(rosidl_runtime_cpp::MessageInitialization _init = rosidl_runtime_cpp::MessageInitialization::ALL)
 @[if init_list]@

--- a/rosidl_generator_cpp/rosidl_generator_cpp/__init__.py
+++ b/rosidl_generator_cpp/rosidl_generator_cpp/__init__.py
@@ -56,7 +56,7 @@ def prefix_with_bom_if_necessary(content: str) -> str:
 MSG_TYPE_TO_CPP = {
     'boolean': 'bool',
     'octet': 'unsigned char',  # TODO change to std::byte with C++17
-    'char': 'unsigned char',
+    'char': 'unsigned char',  # TODO change to char8_t with C++20
     'wchar': 'char16_t',
     'float': 'float',
     'double': 'double',
@@ -197,10 +197,14 @@ def primitive_value_to_cpp(type_, value):
         return 'true' if value else 'false'
 
     if type_.typename in [
+        'char', 'octet'
+    ]:
+        return f'static_cast<unsigned char>({value})'
+
+    if type_.typename in [
         'short', 'unsigned short',
-        'char', 'wchar',
+        'wchar',
         'double', 'long double',
-        'octet',
         'int8', 'uint8',
         'int16', 'uint16',
     ]:
@@ -343,3 +347,44 @@ def create_init_alloc_and_member_lists(message):
                 member_list.append(commonset)
 
     return init_list, alloc_list, member_list
+
+
+def generate_default_string(membset: list) -> list[str]:
+    strlist: list[str] = []
+    for member in membset.members:
+        if member.default_value is not None:
+            if member.num_prealloc > 0:
+                strlist.append('this->%s.resize(%d);' % (member.name, member.num_prealloc))
+            if isinstance(member.default_value, list):
+                if all(v == member.default_value[0] for v in member.default_value):
+                    strlist.append(
+                        'std::fill(this->%s.begin(), this->%s.end(), %s);' % (
+                            member.name,
+                            member.name,
+                            member.default_value[0]
+                            )
+                        )
+                else:
+                    for index, val in enumerate(member.default_value):
+                        strlist.append('this->%s[%d] = %s;' % (member.name, index, val))
+            else:
+                strlist.append('this->%s = %s;' % (member.name, member.default_value))
+
+    return strlist
+
+
+def generate_zero_string(membset: list, fill_args: str) -> list[str]:
+    strlist: list[str] = []
+    for member in membset.members:
+        if isinstance(member.zero_value, list):
+            if member.num_prealloc > 0:
+                strlist.append('this->%s.resize(%d);' % (member.name, member.num_prealloc))
+            if member.zero_need_array_override:
+                strlist.append('this->%s.fill(%s{%s});' % (
+                    member.name, msg_type_only_to_cpp(member.type), fill_args))
+            else:
+                strlist.append('std::fill(this->%s.begin(), this->%s.end(), %s);' % (
+                    member.name, member.name, member.zero_value[0]))
+        else:
+            strlist.append('this->%s = %s;' % (member.name, member.zero_value))
+    return strlist

